File: apps/slate-mac/Sources/SlateMac/Canvas/CanvasAnnouncer.swift
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

import AppKit
import Foundation

// MARK: - CanvasVerbosity (FFI enum + Swift-side niceties)
//
// `CanvasVerbosity` is now the FFI-generated enum in
// `slate_uniffi.swift` (W6-1 PR 0a moved the canvas grammar to core,
// where verbosity is a PARAMETER on the two families whose template
// varies — contracts doc 0a-5). The host used to declare its own
// three-case copy; two same-named enums in one module do not compile,
// and the host copy had nothing left to do once core owned the
// matrix. This file adds the Swift-side niceties on top, exactly like
// `MathPrefs.swift` does for `MathVerbosity`:
// - `CaseIterable` for the Settings picker,
// - `title` for its labels,
// - `Codable` over a stable persistence tag for `PreferencesStore`.
//
// **Persistence-tag stability — DO NOT RENAME.** The literals in
// `persistenceTag` are written into users' UserDefaults through
// `PreferencesStore`; they are the same three strings the previous
// `String`-raw-valued host enum encoded, so stored prefs decode
// unchanged. Renaming one silently resets the user's choice.

extension CanvasVerbosity: Codable, CaseIterable {
    public static var allCases: [CanvasVerbosity] {
        [.terse, .standard, .verbose]
    }

    /// Settings picker label (t0 §1.2 level names).
    var title: String {
        switch self {
        case .terse: return "Terse"
        case .standard: return "Standard"
        case .verbose: return "Verbose"
        }
    }

    private var persistenceTag: String {
        switch self {
        case .terse: return "terse"
        case .standard: return "standard"
        case .verbose: return "verbose"
        }
    }

    public init(from decoder: Decoder) throws {
        let value = try decoder.singleValueContainer().decode(String.self)
        switch value {
        case "terse": self = .terse
        case "standard": self = .standard
        case "verbose": self = .verbose
        default:
            throw DecodingError.dataCorruptedError(
                in: try decoder.singleValueContainer(),
                debugDescription: "Unknown CanvasVerbosity: \(value)"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(persistenceTag)
    }
}

/// Canvas announcement preferences (persisted blob).
struct CanvasPrefs: Codable, Equatable {
    var verbosity: CanvasVerbosity = .standard
}

/// The one announcement funnel for every canvas surface (#518, DoD §H).
///
/// Since W6-1 PR 0a the GRAMMAR is core's: callers hand over a typed
/// `CanvasA11yEvent` from `slate_core::a11y` and this class renders it
/// through the FFI, so the spoken text AND the priority both come from
/// the vocabulary — never composed here, never re-classified here.
/// What stays host-side is exactly what a pure render cannot own: the
/// ~200 ms same-class coalescing window (t0 §1.5 — a render has no
/// clock) and the persisted, live-switchable verbosity preference that
/// the two verbosity-varying families take as a parameter.
///
/// No canvas code calls `postAccessibilityAnnouncement` directly —
/// DoD §H, enforced by
/// `CanvasAnnouncerTests.testNoDirectAnnouncementsUnderCanvas`.
@MainActor
final class CanvasAnnouncer: ObservableObject {
    /// Live-switchable verbosity (§1.2); persisted by the owner and
    /// passed per event (core holds no "current verbosity").
    @Published var verbosity: CanvasVerbosity

    /// Event classes for coalescing (§1.5): same-class events within
    /// the window collapse to the latest. The MEMBERSHIP is pinned
    /// core-side in one doc comment on the canvas family so both hosts
    /// copy one list (contracts doc 0a-8); only the timing is ours.
    private enum EventClass: Hashable {
        case navigation
        case filter
    }

    private let post: (String, NSAccessibilityPriorityLevel) -> Void
    private let coalesceWindow: TimeInterval
    private var pending:
        [EventClass: (work: DispatchWorkItem, text: String, priority: NSAccessibilityPriorityLevel)] = [:]

    /// The seam is `(text, priority)` rather than an event because the
    /// coalescer holds a RENDERED line: the window's winner is decided
    /// after the render, and the loser is dropped without being spoken.
    /// The default posts through the poster layer — never the string
    /// primitive, which belongs to `AnnouncementPosting.swift` alone
    /// (`A11yResidueCensusTests.testNoInteractionSiteCallsTheStringPrimitiveDirectly`).
    init(
        verbosity: CanvasVerbosity = .standard,
        coalesceWindow: TimeInterval = 0.2,
        post: @escaping (String, NSAccessibilityPriorityLevel) -> Void = { text, priority in
            let level: AnnouncementPriority = priority == .high ? .high : .medium
            AppKitAnnouncementPoster().post(text, priority: level)
        }
    ) {
        self.verbosity = verbosity
        self.coalesceWindow = coalesceWindow
        self.post = post
    }

    /// The only announcement API canvas code may use.
    func announce(_ event: CanvasA11yEvent) {
        emit(.canvas(event: event), coalescing: Self.coalescingClass(of: event))
    }

    /// Relay a core event that is NOT canvas vocabulary — the shared
    /// data grid raises its own sort/filter events, and the canvas
    /// table must pass their rendered text AND priority through rather
    /// than re-wrapping them at a priority of its own choosing.
    func relay(_ event: A11yEvent) {
        emit(event, coalescing: nil)
    }

    private func emit(_ event: A11yEvent, coalescing eventClass: EventClass?) {
        let rendered = a11yRender(event: event)
        guard !rendered.text.isEmpty else { return }
        let priority = AnnouncementPriority(rendered.priority)
        if let eventClass {
            debounce(eventClass, text: rendered.text, priority: priority.nsPriority)
            return
        }
        // §1.5: an assertive event supersedes queued navigation
        // context — re-derivable by moving again — so both pending
        // classes are cancelled and DROPPED, never posted.
        if case .high = priority { flushAllPending() }
        post(rendered.text, priority.nsPriority)
    }

    /// The coalescing class of a canvas event, per the list pinned on
    /// the canvas family core-side (contracts doc 0a-8): `navigation`
    /// for movement and transient geometry, `filter` for the two
    /// filter events, immediate for everything else.
    private static func coalescingClass(of event: CanvasA11yEvent) -> EventClass? {
        switch event {
        case .canvasMovedTo, .canvasGroupEntered, .canvasGroupLeft,
            .canvasConnectionTraversed, .canvasMoveRelative, .canvasResizeGeometry:
            return .navigation
        case .canvasFilterCount, .canvasFilterCleared:
            return .filter
        default:
            return nil
        }
    }

    // MARK: Coalescing (§1.5)

    private func debounce(
        _ eventClass: EventClass, text: String, priority: NSAccessibilityPriorityLevel
    ) {
        pending[eventClass]?.work.cancel()
        let work = DispatchWorkItem { [weak self] in
            guard let self else { return }
            guard let entry = self.pending.removeValue(forKey: eventClass) else { return }
            self.post(entry.text, entry.priority)
        }
        pending[eventClass] = (work, text, priority)
        DispatchQueue.main.asyncAfter(deadline: .now() + coalesceWindow, execute: work)
    }

    /// Errors must not be preceded by a stale queued navigation line
    /// (§1.5: the error supersedes; navigation context is re-derivable
    /// by moving again).
    private func flushAllPending() {
        for (_, entry) in pending { entry.work.cancel() }
        pending = [:]
    }

    /// Test hook: emit pending debounced posts NOW (deterministic tests
    /// without wall-clock waits).
    func flushForTests() {
        let items = pending
        pending = [:]
        for (_, entry) in items {
            entry.work.cancel()
            post(entry.text, entry.priority)
        }
    }
}

File: apps/slate-windows/src/SlateWindows/Canvas/CanvasAnnouncer.cs
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

using System.Diagnostics;
using System.Windows.Threading;
using uniffi.slate_uniffi;

namespace SlateWindows.Canvas;

/// <summary>
/// W6-1 PR A (#745), contract A5 (R-C/R-G): the ONE announcement funnel
/// for every canvas surface — the mac <c>CanvasAnnouncer</c> twin.
///
/// The GRAMMAR is core's. Callers hand over a typed
/// <c>CanvasA11yEvent</c> from the PR 0a vocabulary; this class wraps
/// it in <c>A11yEvent.Canvas</c>, renders it through the FFI, and posts
/// the rendered pair — so the spoken text AND the priority both come
/// from the vocabulary, never composed here and never re-classified
/// here. What stays host-side is exactly what a pure render cannot own:
/// the ~200 ms same-class coalescing window (t0 §1.5 — a render has no
/// clock).
///
/// No canvas code posts on its own: <c>CanvasAnnouncerCensus</c> parses
/// every file under <c>Canvas/</c> and fails on any dispatcher
/// reference, any <c>A11yRender</c> call and any <c>HostComposed</c>
/// construction. This file is the single exempt one.
/// </summary>
internal sealed class CanvasAnnouncer
{
    /// <summary>
    /// The coalescing classes (t0 §1.5). MEMBERSHIP is pinned core-side
    /// in one doc comment on the canvas family so both hosts copy one
    /// list (contract 0a-8); only the timing is ours.
    /// </summary>
    private enum EventClass
    {
        Navigation,
        Filter,
    }

    /// <summary>The mac cadence: 200 ms latest-wins, per class.</summary>
    internal static readonly TimeSpan DefaultWindow = TimeSpan.FromMilliseconds(200);

    private readonly Action<RenderedAnnouncement> _post;
    private readonly Dictionary<EventClass, PendingLine> _pending = [];
    private readonly TimeSpan _window;

    /// <summary>
    /// The dispatcher the coalescing timers run on, captured at
    /// CONSTRUCTION rather than at first use.
    /// </summary>
    /// <remarks>
    /// A <see cref="DispatcherTimer"/> binds to
    /// <c>Dispatcher.CurrentDispatcher</c> of whatever thread creates
    /// it, and these are created lazily on first debounce. The
    /// announcer is built on the UI thread by the canvas registry, but
    /// a first navigation announcement arriving from a scheduler body
    /// would otherwise have created a timer on a POOL thread — whose
    /// dispatcher nothing ever pumps, so the queued line would never
    /// fire and the move would simply never be spoken. Binding the
    /// timers to this dispatcher makes the hazard unreachable; the
    /// assert below makes a caller that is off it loud in Debug rather
    /// than merely survivable.
    /// </remarks>
    private readonly Dispatcher _dispatcher;
    private bool _isShutDown;

    /// <summary>
    /// The seam is a RENDERED line rather than an event because the
    /// window's winner is decided after the render and the loser is
    /// dropped without ever being spoken — the mac announcer's
    /// <c>(text, priority)</c> closure, for the same reason. The
    /// default posts through the canonical dispatcher.
    /// </summary>
    public CanvasAnnouncer(
        Action<RenderedAnnouncement> post,
        TimeSpan? coalesceWindow = null)
    {
        ArgumentNullException.ThrowIfNull(post);
        _post = post;
        _window = coalesceWindow ?? DefaultWindow;
        _dispatcher = Dispatcher.CurrentDispatcher;
    }

    /// <summary>The only announcement API canvas code may use.</summary>
    public void Announce(CanvasA11yEvent @event)
    {
        ArgumentNullException.ThrowIfNull(@event);
        Emit(new A11yEvent.Canvas(@event), CoalescingClassOf(@event));
    }

    /// <summary>
    /// Relay a core event that is NOT canvas vocabulary — the shared
    /// <c>AccessibleDataGrid</c> raises its own sort/filter events
    /// (PR B swaps its <c>Announce</c> seam onto this funnel), and the
    /// canvas must pass their rendered text AND priority through rather
    /// than re-wrapping them at a priority of its own choosing.
    /// </summary>
    public void Relay(A11yEvent @event)
    {
        ArgumentNullException.ThrowIfNull(@event);
        Emit(@event, null);
    }

    /// <summary>
    /// Render a LABEL-class event (0a-13): the empty-state onboarding
    /// copy and the degraded banner are composed by core from core's
    /// data, and the surface shows the SAME render the announcement
    /// speaks so the two cannot drift (the mac CD-3 precedent). Not an
    /// announcement — nothing is posted.
    /// </summary>
    public static string RenderLabel(CanvasA11yEvent @event)
    {
        ArgumentNullException.ThrowIfNull(@event);
        return SlateUniffiMethods.A11yRender(new A11yEvent.Canvas(@event)).Text;
    }

    /// <summary>
    /// Retirement: cancel every pending line and refuse anything later
    /// (contract A5).
    /// </summary>
    /// <remarks>
    /// A coalesced line is a timer holding a rendered string. Without
    /// this, closing the last tab on a canvas the user had just moved
    /// around in left a navigation line queued on a document that no
    /// longer exists, and it fired ~200 ms later — the shell speaking
    /// about a surface that is gone. The pending lines are DROPPED, not
    /// flushed: the reason to say them (the user is reading that canvas)
    /// stopped being true.
    /// </remarks>
    internal void Shutdown()
    {
        _isShutDown = true;
        DropAllPending();
    }

    /// <summary>
    /// How many posts this announcer has REFUSED since retirement.
    /// </summary>
    /// <remarks>
    /// The refusal itself is a `Debug.Fail`, which means the guard only
    /// speaks in a developer's build — and CI runs Release, so a shipped
    /// path posting after retirement was invisible to the gate for six
    /// rounds. `Commit`'s confirmation was doing exactly that. A counter
    /// makes "retirement composes nothing" assertable in BOTH
    /// configurations, which is what a guard that matters needs.
    /// </remarks>
    internal int RefusedAfterShutdownForTests { get; private set; }

    private void Emit(A11yEvent @event, EventClass? eventClass)
    {
        if (_isShutDown)
        {
            // Not merely dropped: a post after retirement means some
            // path outlived the document that owns it, and silence would
            // hide that.
            RefusedAfterShutdownForTests++;
            Debug.Fail(
                "CanvasAnnouncer posted after Shutdown: the document it belongs "
                + "to has been retired, so nothing it says can be about a "
                + "surface the user can see.");
            return;
        }
        Debug.Assert(
            _dispatcher.CheckAccess(),
            "CanvasAnnouncer must be driven from the thread it was built on: its "
            + "coalescing timers are bound to that dispatcher, and the publishes "
            + "it feeds are UI state.");
        RenderedAnnouncement rendered = SlateUniffiMethods.A11yRender(@event);
        if (rendered.Text.Length == 0)
        {
            // No canvas template renders empty, so this arm is
            // defensive (mac's `guard !rendered.text.isEmpty` twin) —
            // and a silent drop is the worst possible way to learn that
            // a template started rendering nothing, because the symptom
            // is an announcement that simply does not happen.
            Debug.Fail(
                $"the vocabulary rendered an empty string for {@event.GetType().Name}; "
                + "an announcement was dropped silently.");
            return;
        }
        if (eventClass is { } coalesced)
        {
            Debounce(coalesced, rendered);
            return;
        }
        // t0 §1.5: an assertive event supersedes queued navigation
        // context — re-derivable by moving again — so both pending
        // classes are cancelled and DROPPED, never posted.
        if (rendered.Priority == A11yPriority.High)
        {
            DropAllPending();
        }
        _post(rendered);
    }

    /// <summary>
    /// The coalescing class of a canvas event, per the list pinned on
    /// the canvas family core-side (contract 0a-8): <c>navigation</c>
    /// for movement and transient geometry, <c>filter</c> for the two
    /// filter events, immediate for everything else.
    /// </summary>
    private static EventClass? CoalescingClassOf(CanvasA11yEvent @event) => @event switch
    {
        CanvasA11yEvent.CanvasMovedTo
            or CanvasA11yEvent.CanvasGroupEntered
            or CanvasA11yEvent.CanvasGroupLeft
            or CanvasA11yEvent.CanvasConnectionTraversed
            or CanvasA11yEvent.CanvasMoveRelative
            or CanvasA11yEvent.CanvasResizeGeometry => EventClass.Navigation,
        CanvasA11yEvent.CanvasFilterCount
            or CanvasA11yEvent.CanvasFilterCleared => EventClass.Filter,
        _ => null,
    };

    private void Debounce(EventClass eventClass, RenderedAnnouncement rendered)
    {
        if (!_pending.TryGetValue(eventClass, out PendingLine? line))
        {
            line = new PendingLine(_window, _dispatcher, () => Fire(eventClass));
            _pending[eventClass] = line;
        }
        line.Restart(rendered);
    }

    private void Fire(EventClass eventClass)
    {
        if (!_pending.TryGetValue(eventClass, out PendingLine? line))
        {
            return;
        }
        RenderedAnnouncement? rendered = line.Take();
        if (rendered is not null)
        {
            _post(rendered);
        }
    }

    /// <summary>Errors must not be preceded by a stale queued
    /// navigation line (t0 §1.5).</summary>
    private void DropAllPending()
    {
        foreach (PendingLine line in _pending.Values)
        {
            _ = line.Take();
        }
    }

    /// <summary>Test hook: emit every pending debounced line NOW, so a
    /// fact is deterministic without a wall-clock wait (the mac
    /// <c>flushForTests</c> twin).</summary>
    internal void FlushForTests()
    {
        foreach (EventClass eventClass in _pending.Keys.ToList())
        {
            Fire(eventClass);
        }
    }

    /// <summary>One class's queued line and its timer. The timer is
    /// stopped whenever the line is taken, so a fired-and-drained class
    /// cannot re-post the same text.</summary>
    private sealed class PendingLine
    {
        private readonly DispatcherTimer _timer;
        private RenderedAnnouncement? _rendered;

        internal PendingLine(TimeSpan window, Dispatcher dispatcher, Action onElapsed)
        {
            // The four-argument ctor binds the timer to the GIVEN
            // dispatcher; the parameterless one would bind it to
            // whatever thread got here first.
            _timer = new DispatcherTimer(DispatcherPriority.Normal, dispatcher)
            {
                Interval = window,
            };
            _timer.Tick += (_, _) =>
            {
                _timer.Stop();
                onElapsed();
            };
        }

        internal void Restart(RenderedAnnouncement rendered)
        {
            _rendered = rendered;
            _timer.Stop();
            _timer.Start();
        }

        internal RenderedAnnouncement? Take()
        {
            _timer.Stop();
            RenderedAnnouncement? taken = _rendered;
            _rendered = null;
            return taken;
        }
    }
}
